sdk/python/kfp/v2/components/components.py

```diff
@@ -0,0 +1,667 @@
+      elif isinstance(argument_value, structures.TaskOutputArgument):
+        reference_type = argument_value.task_output.type
+      else:
+        reference_type = None
+
+      if reference_type and input_type and reference_type != input_type:
+        _react_to_incompatible_reference_type(input_type, reference_type,
+                                              input_name)
+
+      task_arguments[input_name] = argument_value
+    else:
+      # argument_value is a constant value
+      serialized_argument_value = serialize_value(argument_value, input_type)
+      task_arguments[input_name] = serialized_argument_value
+
+  task = structures.TaskSpec(
+      component_ref=component_ref,
+      arguments=task_arguments,
+  )
+  task._init_outputs()
+
+  return task
+
+
+_default_container_task_constructor = _create_task_spec_from_component_and_arguments
+
+# Holds the function that constructs a task object based on ComponentSpec, arguments and ComponentReference.
+# Framework authors can override this constructor function to construct different framework-specific task-like objects.
+# The task object should have the task.outputs dictionary with keys corresponding to the ComponentSpec outputs.
+# The default constructor creates and instance of the TaskSpec class.
+_container_task_constructor = _default_container_task_constructor
+
+_always_expand_graph_components = False
+
+
+def _create_task_object_from_component_and_arguments(
+    component_spec: structures.ComponentSpec,
+    arguments: Mapping[str, Any],
+    component_ref: structures.ComponentReference = None,
+    **kwargs):
+  """Creates a task object from component and argument.
+
+    Unlike _container_task_constructor, handles the graph components as well.
+    """
+  if (isinstance(
+      component_spec.implementation, structures.GraphImplementation
+  ) and (
+      # When the container task constructor is not overriden, we just construct TaskSpec for both container and graph tasks.
+      # If the container task constructor is overriden, we should expand the graph components so that the override is called for all sub-tasks.
+      _container_task_constructor != _default_container_task_constructor or
+      _always_expand_graph_components)):
+    return _resolve_graph_task(
+        component_spec=component_spec,
+        arguments=arguments,
+        component_ref=component_ref,
+        **kwargs,
+    )
+
+  task = _container_task_constructor(
+      component_spec=component_spec,
+      arguments=arguments,
+      component_ref=component_ref,
+      **kwargs,
+  )
+
+  return task
+
+
+class _DefaultValue:
+
+  def __init__(self, value):
+    self.value = value
+
+  def __repr__(self):
+    return repr(self.value)
+
+
+#TODO: Refactor the function to make it shorter
+def _create_task_factory_from_component_spec(
+    component_spec: structures.ComponentSpec,
+    component_filename=None,
+    component_ref: structures.ComponentReference = None):
```

Comment:
Okay to leave it as-is, so treat this comment as non-blocking:

component_ref will be eventually used as a k8s pod annotation when running in Argo, which is not yet implemented. We might not need this complexity here.